### PR TITLE
ORA1761 New “strict” query that returns latest inferences and losses when both are fully available

### PR DESCRIPTION
**Pull request by br4e**
This includes both inferences and losses

### CODE REVIEW

File: x/emissions/proto/emissions/v1/query.proto [hunk 1]
```diff
@@ -170,8 +170,13 @@ service Query {
...
   }
 
-  rpc GetLatestNetworkInference(QueryLatestNetworkInferencesAtBlockRequest)
-      returns (QueryLatestNetworkInferencesAtBlockResponse) {
+  rpc GetLatestNetworkInference(QueryLatestNetworkInferencesRequest) returns (QueryLatestNetworkInferencesResponse) {
+    option (cosmos.query.v1.module_query_safe) = true;
+    option (google.api.http).get = "/emissions/v1/network_inferences/{topic_id}";
+  }
+
+  rpc GetLatestNetworkInferenceWithLosses(QueryLatestNetworkInferencesRequest)
```
Comment (guilherme-brandao) [inline x/emissions/proto/emissions/v1/query.proto:178]: I'd name this something like`GetLatestAvailableNetworkInference`

File: x/emissions/keeper/queryserver/query_server_inferences.go [hunk 2]
```diff
@@ -80,10 +83,76 @@ func (qs queryServer) GetLatestNetworkInference(
...
+	if err != nil {
+		return nil, err
+	}
+
+	previousLossBlockHeight := inferenceBlockHeight - topic.EpochLength
+
+	for true {
+		if previousLossBlockHeight < 0 {
+			return nil, status.Errorf(codes.NotFound, "losses not yet available for topic %v", req.TopicId)
+		}
```
Comment (guilherme-brandao) [inline x/emissions/keeper/queryserver/query_server_inferences.go:118]: We can have this case and also a case where losses are available but there isn't a subsequent inference. Since this is a 'strict' query, I think we should also cover this case.

**Review by guilherme-brandao — CHANGES_REQUESTED** [review]

File: x/emissions/keeper/queryserver/query_server_inferences.go (hunk 2 above)
Comment (br4e) [reply x/emissions/keeper/queryserver/query_server_inferences.go:135]: Hmm this should be handled with the line `if len(inferences.Inferences) > 0 {` 

right?

**Comment by br4e** [comment]
Also are we worried about a possible infinite loop? Should we limit the iterations to 10 epochs back or something?

File: x/emissions/keeper/queryserver/query_server_inferences.go (hunk 2 above)
Comment (br4e) [reply x/emissions/keeper/queryserver/query_server_inferences.go:135]: ah with some more tests I see the edge case. Working on a fix

Comment (br4e) [reply x/emissions/keeper/queryserver/query_server_inferences.go:144]: fixed!

File: x/emissions/keeper/queryserver/query_server_inferences.go [hunk 3]
```diff
@@ -97,10 +100,81 @@ func (qs queryServer) GetLatestNetworkInference(
...
+) (
+	*types.QueryLatestNetworkInferencesResponse,
+	error,
+) {
+
+	// Find the latest inference block height
+	_, inferenceBlockHeight, err := qs.k.GetLatestTopicInferences(ctx, req.TopicId)
+	if err != nil {
+		return nil, err
+	}
```
Comment (guilherme-brandao) [inline x/emissions/keeper/queryserver/query_server_inferences.go:125]: You can also add a check here if the blockHeight is equal to zero like in the other queries

Comment (br4e) [reply x/emissions/keeper/queryserver/query_server_inferences.go:125]: updated

**Comment by br4e** [comment]
I think default page limit makes sense, it's kind of like looking through that many "pages" of inferences before giving up so we don't have something like an infinite loop.

File: x/emissions/keeper/queryserver/query_server_inferences.go [hunk 4]
```diff
@@ -97,10 +100,84 @@ func (qs queryServer) GetLatestNetworkInference(
...
+			// check for losses. If they exist, then break so we can query
+			_, err = qs.k.GetNetworkLossBundleAtBlock(ctx, req.TopicId, previousLossBlockHeight)
+			if err == nil {
+				break
+			}
+		}
+
+		inferenceBlockHeight -= topic.EpochLength
+		previousLossBlockHeight -= topic.EpochLength
+	}
```
Comment (kpeluso) [inline x/emissions/keeper/queryserver/query_server_inferences.go:160]: 1. previous block height won't strictly be at that time e.g. if the topic was inactive for a period
2. this is a very inefficient way to discover the right block. we should find an appropriate block and then use functions (GetInferencesAtBlock, GetNetworkLossBundleAtBlock) for their intended purpose given the appropriate blocks

**Review by kpeluso — CHANGES_REQUESTED** [review]

File: x/emissions/keeper/queryserver/query_server_inferences.go (hunk 4 above)
Comment (br4e) [reply x/emissions/keeper/queryserver/query_server_inferences.go:160]: Do you have an alternative method in mind for the second point?

**Comment by br4e** [comment]
Ok, this is updated and ready for re-review. I recorded the relevant information so it can be used to determine the latest blocks for reputer and works payloads

**Review by kpeluso — APPROVED** [review]
